### DISCUSSION TITLE
Automatic detection of attributes for ransack allowlist

### DISCUSSION THREAD

**Discussion by taras-viiatyk-yalantis** [activeadmin/activeadmin#8224] [Ideas]
Hi everyone! 

Before updating to version 3, the following code just worked fine with my Post model:
```ruby
ActiveAdmin.register Post do
  # ...
  filter :name
  filter :description
  # ...
end
```

After updating to version 3, this code stopped working with the following error:
```
Ransack needs Post associations explicitly allowlisted as
searchable. Define a `ransackable_attributes` class method in your `Post`
model, watching out for items you DON'T want searchable (for
example, `encrypted_password`, `password_reset_token`, `owner` or
other sensitive information). You can use the following as a base:

class Post < ApplicationRecord

  # ...

  def self.ransackable_attributes(auth_object = nil)
    ["id", "name", "description", "created_at", "updated_at"]
  end

  # ...

end
```

I had to put the code similar to the following inside each of the filterable registered models:
```ruby
class Post < ApplicationRecord
  def self.ransackable_attributes(_auth_object = nil)
    ["name" "description"]
  end
end
```
I have around 50 such models. The process of putting these codes inside each model is pretty long and routine. A similar situation happens with ransackable_associations.

Is there any way to autodetect which attributes should be put to ransack allowlist based on defined filters? It seems to me obvious that if the filter is defined for some attribute, then that attribute should be put into the allowlist of ransack. Moreover, it's pretty easy to forget to remove an attribute from the allowlist once the corresponding filter has been removed for any reason, so it would be great to automatically delete an attribute from the allowlist once a filter is deleted.

**Comment by amkisko** [comment]
Quick temporary solution is to put this to ApplicationRecord:
```
  def self.ransackable_associations(*, **) = reflections.keys
  def self.ransackable_attributes(*, **) = attribute_names
```

**Reply by taras-viiatyk-yalantis** [reply]
I decided to not use this temporary solution since it basically reverts the changes related to the security made in the ransack library.

**Comment by caifara** [comment]
I decided to temporarily go with @amkisko's solution with some obvious overrides like user encrypted passwords (and warnings added).

I started to look into some implementation to automate this, but soon realised that it's not because a filter is added to some AA resource, that every user may use ransack on that attribute. The AA resource or the filter may not be accessible to that person. I suppose that any official support would need more nuance.

I think I'll have to go through all our models eventually to add a whitelist (once we're able to upgrade to v4 which passes an auth_object).

**Reply by javierjulio** [reply]
@caifara if it helps for your last comment, [that change was a one-liner](https://github.com/activeadmin/activeadmin/pull/8143/files) so should be ok to monkey patch in on v3 and remove on v4. We weren't sure of the safety of the change so its why it was included in a major release.
